Feature/adding and removing watchers

tilt opened the pull request:
https://www.openproject.org/work_packages/13953

Commit:
Added failing specs for API watcher resource

tilt committed:
Add watchers#create API end point

tilt committed:
Render API error response when watcher is invalid

tilt committed:
Let API render status 200 if a user is already a watcher

tilt committed:
Add authorization to API watchers#create

tilt committed:
Implement API watchers#delete end point

myabc commented on lib/api/v3/work_packages/watchers_api.rb:
```diff
@@ -0,0 +1,54 @@
+
+            user = User.find params[:user_id]
+
+            if @work_package.watcher_users.include?(user)
```
There is a lot of code here that we might want to move elsewhere. In other words, our Grape code should not end up looking like the Rails Controllers we're replacing.

tilt replied:
Any suggestions for restructuring it?

myabc replied:
https://www.youtube.com/watch?v=tg5RFeSfBM4 around 45 minutes in.

FWIW DHH doesn't like this approach (calling it ["Test Induced Design Damage"](https://gist.github.com/dhh/4849a20d2ba89b34b201)).

myabc commented on spec/api/watcher_resource_spec.rb:
```diff
@@ -0,0 +1,131 @@
+      let(:current_user) { FactoryGirl.create :user,  member_in_project: project, member_through_role: add_watchers_role }
+
+      it 'should respond with 201' do
+        expect(subject.status).to eq(201)
```
if you're going to use named subject, I think `response.status` is clearer than `subject.status`.

myabc commented:
Looks pretty good so far.

tilt committed:
Remove duplicate code block

tilt committed:
Add watch and unwatch links to wp resource (WIP)


TODO
	- fix specs
	- prettify code

Commit:
SASS change for watch / unwatch WP

tilt committed:
Allow users allowed to view work packages to (un-)watch via API v3 (WIP)


TODO better code style

tilt committed:
Services for watcher creation / removal

tilt committed:
Use type coercion to convert id params

tilt committed:
Merge branch 'feature/adding-and-removing-watchers' of github.com:opf/openproject into feature/adding-and-removing-watchers

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -68,10 +70,27 @@ def initialize(options = {}, *expand)
           {
               href: "#{root_url}/api/v3/users/#{represented.work_package.assigned_to.id}",
-              title: "#{represented.work_package.assigned_to.name} - #{represented.work_package.assigned_to.login}"
+              title: "#{represented.work_package.assigned_to.name} - #{represented.work_package.assigned_to.login}",
```
the ", " at the end needs to be removed

tilt replied:
this was an accident

tilt committed:
Remove accidentally added comma

tilt committed:
Display delete button next to watchers in details tab

Commit:
Watch / unwatch work packages

Commit:
Merge branch 'feature/adding-and-removing-watchers' of github.com:opf/openproject into feature/adding-and-removing-watchers

tilt committed:
Merge branch 'feature/adding-and-removing-watchers' of github.com:opf/openproject into feature/adding-and-removing-watchers

tilt committed:
Avoid error when representer is initialized without current user

tilt committed:
Spec presence of (un-)watch link in representer spec

tilt committed:
Extend work package representer by add / remove watcher links (WIP)


FIXME the delete link requires the user id in its href

tilt committed:
De-implement delete watcher link of wp representer

tilt committed:
Start enabling (un-)watching/-starring of work packages (WIP)


FIXME watcher deletion

Status 204 means empty response bodies on the API (grape) side, whereas the hyperagent cannot deal with those on the client side.

Suggestions: Use $http, find out how to intercept and patch the response or make hyper agent more tolerant

Commit:
Add link to remove watcher to the user representer

Commit:
Using hyperagent version which swallows up empty responses from delete requests.

Commit:
Merge branch 'feature/adding-and-removing-watchers' of github.com:opf/openproject into feature/adding-and-removing-watchers

tilt committed:
Remove flash when (un-)watching work package, add better error handling

tilt committed:
Wire up remove watcher links

tilt committed:
Deal with work package refreshes and update watchers on watchers tab

tilt committed:
Adapt star icon active state when the user removes herself from the list

Commit:
Added tooltips for watch / unwatch WP

Commit:
Merge branch 'feature/adding-and-removing-watchers' of github.com:opf/openproject into feature/adding-and-removing-watchers

tilt committed:
Remove some code duplication when refreshing work packages

Commit:
Anon. user cannot see link to watch WP

Commit:
Merge branch 'feature/adding-and-removing-watchers' of github.com:opf/openproject into feature/adding-and-removing-watchers

Commit:
Hid star for watching users if user is anon.

tilt committed:
Dry up error handling when refreshing work package on details pane

tilt committed:
Merge branch 'feature/adding-and-removing-watchers' of github.com:opf/openproject into feature/adding-and-removing-watchers

tilt committed:
Update delete watcher locales

tilt committed:
Fix karma

tilt committed:
Merge branch 'dev' into feature/adding-and-removing-watchers

Commit:
WIP create available watchers endpoint

Commit:
Added build_representer helper

Commit:
WIP Basic wire-up of Watchers API with angular

tilt committed:
Fix karma (mock work package links)

tilt committed:
Fix work package representer spec

tilt committed:
Add drop down with available watchers to watchers tab (WIP)


TODO to be replaced with a generalized widget for auxiliary resources

tilt committed:
Add placeholder for adding watchers

tilt committed:
Merge branch 'dev' into feature/adding-and-removing-watchers

Conflicts:
	app/assets/javascripts/angular/work_packages/controllers/work-package-details-controller.js
	karma/tests/controllers/work-package-details-controller-test.js
	lib/api/v3/work_packages/work_package_representer.rb

tilt committed:
Merge branch 'dev' into feature/adding-and-removing-watchers

Conflicts:
	app/assets/javascripts/angular/work_packages/controllers/work-package-details-controller.js
	karma/tests/controllers/work-package-details-controller-test.js
	lib/api/v3/work_packages/work_package_representer.rb

tilt committed:
Temporarily filter available watchers on the client-side (WIP)


… until the API does the work.
